libraries/AP_Scripting/examples/ahrs-source-gps-optflow.lua

```diff
@@ -0,0 +1,207 @@
+-- switches between AHRS/EKF sources based on the pilot's source selection switch or using an automatic source selection algorithm
+-- this script is intended to help vehicles automatically switch between GPS and optical flow
+--
+-- configure a forward or downward facing lidar with a range of at least 5m
+-- setup RCx_OPTION = 90 (EKF Pos Source) to select the source (low=GPS, middle=opticalflow, high=Not Used)
+-- setup RCx_OPTION = 300 (Scripting1).  When this switch is pulled high, the source will be automatically selected
+-- SRC_ENABLE = 1 (enable scripting)
+-- setup EK3_SRCn_ parameters so that GPS is the primary source, opticalflow is secondary.
+--     EK3_SRC1_POSXY = 3 (GPS)
+--     EK3_SRC1_VELXY = 3 (GPS)
+--     EK3_SRC1_VELZ  = 3 (GPS)
+--     EK3_SRC1_POSZ  = 1 (Baro)
+--     EK3_SRC1_YAW   = 1 (Compass)
+--     EK3_SRC2_POSXY = 0 (None)
+--     EK3_SRC2_VELXY = 5 (OpticalFlow)
+--     EK3_SRC2_VELZ  = 0 (None)
+--     EK3_SRC2_POSZ  = 1 (Baro)
+--     EK3_SRC2_YAW   = 1 (Compass)
+--     EK3_OPTIONS    = 0 (Do not fuse all velocities)
+--
+-- SCR_USER1 holds the threshold (in meters) for rangefinder altitude (around 15 is a good choice)
+--     if rangefinder distance >= SCR_USER1, source1 (GPS) will be used
+--     if rangefinder distance < SCR_USER1, source2 (optical flow) will be used if innovations are below SRC_USER3 value
+-- SCR_USER2 holds the threshold for GPS speed accuracy (around 0.3 is a good choice)
+-- SCR_USER3 holds the threshold for optical flow quality (about 50 is a good choice)
+-- SCR_USER4 holds the threshold for optical flow innovations (about 0.15 is a good choice)
+--
+-- When the 2nd auxiliary switch (300/Scripting1) is pulled high automatic source selection uses these thresholds:
+
+local rangefinder_rotation = 25     -- check downward (25) facing lidar
+local source_prev = 0               -- previous source, defaults to primary source
+local sw_source_prev = -1           -- previous source switch position
+local sw_auto_pos_prev = -1         -- previous auto source switch position
+local auto_switch = false           -- true when auto switching between sources is active
+local gps_usable_accuracy = 1.0     -- GPS is usable if speed accuracy is at or below this value
+local vote_counter_max = 20         -- when a vote counter reaches this number (i.e. 2sec) source may be switched
+local gps_vs_opticalflow_vote = 0   -- vote counter for GPS vs optical (-20 = GPS, +20 = optical flow)
+local scr_user1_param = Parameter() -- user1 param (rangefinder altitude threshold)
+local scr_user2_param = Parameter() -- user2 param (GPS speed accuracy threshold)
+local scr_user3_param = Parameter() -- user3 param (optical flow quality threshold)
+local scr_user4_param = Parameter() -- user4 param (optical flow innovation threshold)
+
+-- initialise parameters
+assert(scr_user1_param:init('SCR_USER1'), 'ahrs-source.lua: could not find SCR_USER1 parameter')
```

Comment:
This prints via the lua error function so you will already get the file name and line number. 
```suggestion
assert(scr_user1_param:init('SCR_USER1'), 'could not find SCR_USER1 parameter')
```

no bid deal tho, TBH I don't think I have ever seen one of these fail so long as the param name is correct and its not like this param is vehicle dependant.

Reply:
fixed thanks.